lxd/daemon_images.go

```diff
@@ -0,0 +1,265 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime"
+	"mime/multipart"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/lxc/lxd/shared"
+
+	log "gopkg.in/inconshreveable/log15.v2"
+)
+
+// ImageDownload checks if we have that Image Fingerprint else
+// downloads the image from a remote server.
+func (d *Daemon) ImageDownload(
+	server, fp string, secret string, forContainer bool) error {
+
+	if id := dbImageIDGet(d.db, fp); id != -1 {
+		shared.Log.Debug("Image already exists in the db", log.Ctx{"image": fp})
+		// already have it
+		return nil
+	}
+
+	shared.Log.Info(
+		"Image not in the db downloading it",
+		log.Ctx{"image": fp, "server": server})
+
+	// Now check if we already downloading the image
+	d.imagesDownloadingLock.RLock()
+	if lock, ok := d.imagesDownloading[fp]; ok {
+		// We already download the image
+		d.imagesDownloadingLock.RUnlock()
+
+		shared.Log.Info(
+			"Already downloading the image, waiting for it to succeed",
+			log.Ctx{"image": fp})
+
+		// Now we use a little trick
+		// we wait until we can lock the download once
+		// we are able to lock it we return we have it on succeed.
+		lock.Lock()
```

Comment:
I think this might be cleaner using the concurrency primitives go provides, namely channels. I think instead of using a lock, we can just use a channel, and have the goroutine block until the channel is closed. It'll just look a little less awkward.